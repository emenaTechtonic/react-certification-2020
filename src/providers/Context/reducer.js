export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      }
    case 'SET_FAVORITES':
      return {
        ...state,
        favorites: action.payload,
      }

    default:
      return state
  }
}
