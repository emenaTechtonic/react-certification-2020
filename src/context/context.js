import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
