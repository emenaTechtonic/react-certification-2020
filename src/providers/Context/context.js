import React, { useEffect, createContext, useCallback, useReducer, memo } from 'react'
import { useFetch } from '../../utils/hooks/useFetch'
import { storage } from '../../utils/storage'
import { isElementIn, removeElement } from '../../utils/fns'
import {
  AUTH_STORAGE_USER,
  AUTH_STORAGE_KEY,
  AUTH_STORAGE_FAVORITES,
} from '../../utils/constants'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])
  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    if (isAuthenticated) {
      const lastUser = storage.get(AUTH_STORAGE_USER)
      const userAuthenticated = Object(lastUser)

      dispatch({ type: 'SET_USER', payload: userAuthenticated })

      const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
      if (lastFavorites) {
        const favorites = lastFavorites
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
      }
    }
  }, [])

  const saveFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const res = isElementIn(card.id, favorites)
      if (!res) {
        favorites.push(card)
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
        storage.set(AUTH_STORAGE_FAVORITES, favorites)
      }
    }
  }, [])
  const removeFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const newFavorites = removeElement(card.id, favorites)
      dispatch({ type: 'SET_FAVORITES', payload: newFavorites })
      storage.set(AUTH_STORAGE_FAVORITES, newFavorites)
    }
  }, [])

  // return isLoading ? (
  //   <div> Loading...</div>
  // ) : (
  return (
    <Context.Provider value={{ state, dispatch, saveFavoriteCard, removeFavoriteCard }}>
      {children}
    </Context.Provider>
  )
}
export default memo(ContextProvider)
