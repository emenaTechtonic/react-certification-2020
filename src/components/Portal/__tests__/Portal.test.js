import React from 'react'
import renderer from 'react-test-renderer'
import { Context } from '../../../providers/Context/context'
import Portal from '../Portal'

const state = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
}
describe('Testing Portal component', () => {
  test('Renders Portal without crashing', async () => {
    const rendered = renderer.create(
      <Context.Provider value={{ state }}>
        <Portal />
      </Context.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })
})
