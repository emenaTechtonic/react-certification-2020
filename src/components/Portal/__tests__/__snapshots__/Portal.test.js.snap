// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Portal component Renders Portal without crashing 1`] = `null`;
